### ISSUE TITLE
Custom authentication provider (e.g. AWS Cognito User Pools) not working because subdomain parameter not set in grant object configuration

### ISSUE THREAD

**Issue by ralphsomeday** [strapi/strapi#6822]
**Describe the bug**
When trying to create a new custom authentication provider for strapi, I am calling the strapi server base url /connect/new_provider_name then I am getting a dns error from the browser. When looking at the network log I can see a 302 response with a redirection to https://[subdomain]/oauth2/authorize. The subdomain is not being replaced by the link to the Identity provider (in this case AWS Cognito User Pools)

**Steps to reproduce the behavior**

1. Follow the documentation to add a new Custom provider
2. Go to https://strapi_server_url/collect/name_of_your_provider
3. Redirection fails

**Expected behavior**
The redirection from https://strapi_server_url/collect/name_of_your_provider to the identity provider authorize page works

**Code snippets**
The issue seems to be in not having the subdomain property configurable in the grant object configuration in the user-permissions plugin and more precisely in the connect method located in the Auth.js controller (users-permissions > controllers > Auth.js

```javascript
async connect (ctx, next) {
    const grantConfig = await strapi
      .store({
        environment: '',
        type: 'plugin',
        name: 'users-permissions',
        key: 'grant'
      })
      .get()

    const [requestPath] = ctx.request.url.split('?')
    const provider =
      process.platform === 'win32' ? requestPath.split('\\')[2] : requestPath.split('/')[2]

    if (!_.get(grantConfig[provider], 'enabled')) {
      return ctx.badRequest(null, 'This provider is disabled.')
    }
    // Ability to pass OAuth callback dynamically
    grantConfig[provider].callback = _.get(ctx, 'query.callback') || grantConfig[provider].callback
    grantConfig[provider].redirect_uri = `${strapi.config.server.url}/connect/${provider}/callback`
    // Adding the following code seems to solve the issue
    grantConfig[provider].subdomain = strapi.config.custom.cognitoHost

    return grant(grantConfig)(ctx, next)
},
```
Adding the following code in the previous method seems to work:
```javascript
    // Adding the following sentence seems to solve the issue
    grantConfig[provider].subdomain = strapi.config.custom.cognitoHost
```

**System**

- Node.js version: v12
- Strapi version: v3.0.5

**Comment by lauriejim** [comment]
Hello! thank you for this message.
I think it could be a good idea to manage more grant options like `subdomain` in this case.
Feel free to submit a PR, I will appreciate that.
(but think that should not be only for your use case w/ AWS Cognito :) )
Thank you!

**Comment by ralphsomeday** [comment]
@lauriejim PR submitted for your review

**Comment by lauriejim** [comment]
Thank you! I will check that soon as possible :)

**Comment by JiboStore** [comment]
@ralphsomeday did you manage to get the whole flow working?

In my case, the first time I login, it will launch the cognito hosted UI, from there, it will redirect to the specified callback url (${strapi_url}/connect/cognito/callback) with the GET params: id_token, token_type, etc. However, the browser displayed `HTTP_ERROR 400. This Page isn't working.` Tracing the strapi's user-permission plugin, that endpoint will call Auth.js `connect()` function which goes all the way to the end (verified through log) but nothing else happens after that.

The second time logs-in successfully, the cognito hosted UI never pops up anymore, although the same Auth.js `connect()` function gets called and all the way to the end AND THEN Providers.js `connect()` function gets called, which connects and logs in the user!

Anyone has any idea how to debug the difference? That last line of code in Auth.js is `return grant(grantConfig)(ctx, next);` I am not sure what happens inside `grant` nor how to modify their code to trace and debug what is happening.. It could be (and most likely) just a simple config that I missed

**Comment by ralphsomeday** [comment]
@JiboStore the redirect URL to your front app when configuring the provider is the one that should point to your front end app and not to ${strapi_url}/connect/cognito/callback. Your frontend app will retrieve the id_token from the url parameters and use it to call the strapi api endpoint to retrieve the user profile for example or call protected content types api endpoints.

![image](https://user-images.githubusercontent.com/61593456/92998527-7b9c0a00-f51a-11ea-9633-b5a5ffee34d5.png)

**Comment by JiboStore** [comment]
@ralphsomeday that means you must change the user-service (under services/User.js) to retrieve the user from cognito instead? Only that then you can use cognito's token to authenticate with your strapi.

Ideally, I'd like to do that too, the benefit of it is that we'll have all our users in the cognito User-Pool instead of the strapi's db, however I'm just not sure of the amount of work involved and if I am only doing it this way for a specific provider (cognito), then it will break away the rest of the providers, and perhaps will need to maintain our own for the rest of the strapi's version updates.

Otherwise, the default strapi's behavior (I learnt from strapi's code for the Facebook Provider, cmiiw) is that strapi is just using the third party token to retrieve the information, create its own strapi's User (with fields populated with information from the provider) and then issue another jwt token which is used in subsequent strapi requests

**Comment by ralphsomeday** [comment]
@JiboStore if you wish to retrieve the cognito user information you can just use AWS Amplify in your front end with the id_token retrieved from strapi and you should be able to get all the information directly from cognito.

**Comment by ralphsomeday** [comment]
@JiboStore PR is now ready for review by @alexandrebodin and if all good should be merged and maybe available in version 3.1.7
